iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_fifo.inl

```diff
@@ -27,33 +27,47 @@ namespace concurrent
 template <class ValueType, uint64_t Capacity>
 inline bool SpscFifo<ValueType, Capacity>::push(const ValueType& value) noexcept
 {
-    if (is_full())
+    // Memory order relaxed is enough since:
+    // - there is no concurrent access to this method
+    // - the load statement cannot be reordered with writing m_data otherwise it would not compile
+    auto currentWritePos = m_writePos.load(std::memory_order_relaxed);
+    bool isFull = (currentWritePos ==
+                   // Memory order acquire is needed to ensure happens-before relationship
+                   m_readPos.load(std::memory_order_acquire) + Capacity);
+    if (isFull)
     {
         return false;
     }
-    auto currentWritePos = m_write_pos.load(std::memory_order_relaxed);
     m_data[currentWritePos % Capacity] = value;
 
-    // m_write_pos must be increased after writing the new value otherwise
-    // it is possible that the value is read by pop while it is written.
-    // this fifo is a single producer, single consumer fifo therefore
-    // store is allowed.
-    m_write_pos.store(currentWritePos + 1, std::memory_order_release);
+    // SYNC POINT WRITE: m_data
+    // We need to make sure that writing the value happens before incrementing the
+    // m_writePos otherwise the following scenario can happen:
+    // 1. m_writePos is increased (but the value has not been written yet)
+    // 2. Another thread calls pop(): we check if the queue is empty => no
+    // 3. In pop(), when we read a value, a data race can occur when at the same time a value is
+    // written by push. With memory_order_release, this cannot happen as it is guaranteed that
+    // writing the data happens before incrementing m_writePos. Note that the following scenario
+    // can still happen (but this is not a problem):
+    // 1. A value is written (m_writePos hasn't been incremented yet)
+    // 2. Another thread calls pop(): we check if the queue is empty => yes
+    // 3. An element was already stored so we could have popped the element
```

Comment:
I would also rephrase this since it is a bit misleading. Technically the consumer thread cannot read the element, even after the producer thread wrote the data since the producer did not yet synchronized the memory. Stating that the consumer threat could have popped the element is therefore not true. It could only have popped the element if it would have been informed that there is one, everything else is not allowed and leads to a data race.

1. There is an empty queue
2. A push operation is in progress, the value has ben written but 'm_writePos' was not yet advanced
3. The consumer thread performs a pop operation and the check for an empty queue is true resulting in a failed pop
4. The push operation is finished by advancing `m_writePos` and synchronizing the memory
5. The consumer thread missed the chance to pop the element by a blink of an eye